kms/capi/capi.go

```diff
@@ -0,0 +1,754 @@
+	if req.Name == "" {
+		return nil, errors.New("createKeyRequest 'name' cannot be empty")
+	}
+
+	// The MSSC provider allows you to create keys without a certificate attached, but they seem to
+	// be lost if the smartcard is removed, so refuse to create keys as a precaution
+	if k.providerName == ProviderMSSC {
+		return nil, fmt.Errorf("cannot create keys on %s", ProviderMSSC)
+	}
+
+	u, err := uri.ParseWithScheme(Scheme, req.Name)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse URI: %w", err)
+	}
+
+	var containerName string
+	if containerName = u.Get(ContainerNameArg); containerName == "" {
+		// generate a uuid for the container name
+		containerName, err = randutil.UUIDv4()
+		if err != nil {
+			return nil, fmt.Errorf("failed to generate uuid: %w", err)
+		}
+	}
+
+	alg, ok := signatureAlgorithmMapping[req.SignatureAlgorithm]
+	if !ok {
+		return nil, fmt.Errorf("unsupported algorithm %v", req.SignatureAlgorithm)
+	}
+
+	//TODO: check whether RSA keys require legacyKeySpec set to AT_KEYEXCHANGE
+	kh, err := nCryptCreatePersistedKey(k.providerHandle, containerName, alg, 0, 0)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create persisted key: %w", err)
+	}
+
+	defer nCryptFreeObject(kh)
+
+	if alg == "RSA" {
+		err = nCryptSetProperty(kh, NCRYPT_LENGTH_PROPERTY, uint32(req.Bits), 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_LENGTH_PROPERTY: %w", err)
+		}
+	}
+
+	// users can store the key as a machine key by passing in storelocation = machine
+	// 'machine' is the only valid location, otherwise the key is stored as a 'user' key
+	storeLocation := u.Get(StoreLocationArg)
+
+	if storeLocation == "machine" {
+		err = nCryptSetProperty(kh, NCRYPT_KEY_TYPE_PROPERTY, NCRYPT_MACHINE_KEY_FLAG, 0)
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_KEY_TYPE_PROPERTY: %w", err)
+		}
+	} else if storeLocation != "" && storeLocation != "user" {
+		return nil, fmt.Errorf("invalid storeLocation %v", storeLocation)
+	}
+
+	// if supplied, set the smart card pin/or PCP pass
+	pinOrPass := u.Pin()
+
+	//failover to pin set in kms instantiation
+	if pinOrPass == "" {
+		pinOrPass = k.pin
+	}
+
+	// TODO: investigate if there is a similar property for software backed keys
+	if pinOrPass != "" && k.providerName == ProviderMSSC {
+		err = nCryptSetProperty(kh, NCRYPT_PIN_PROPERTY, pinOrPass, 0)
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PIN_PROPERTY: %w", err)
+		}
+	} else if pinOrPass != "" && k.providerName == ProviderMSPCP {
+		pwHash, err := hashPasswordUTF16(pinOrPass) // we have to SHA1 hash over the utf16 string
+
+		if err != nil {
+			return nil, fmt.Errorf("unable to hash pin: %w", err)
+		}
+		err = nCryptSetProperty(kh, NCRYPT_PCP_USAGE_AUTH_PROPERTY, pwHash, 0)
+		if err != nil {
+			return nil, fmt.Errorf("unable to set key NCRYPT_PIN_PROPERTY: %w", err)
+		}
+	}
```

Comment:
Is there a missing else here?

Reply:
@maraino depends if you want the key creation to fail when the user provides a pin but specifies the MSKSP (ie software) provider?

Reply:
Well, I trust your judgment here. We can allow the ProviderMSKSP with a PIN even if it never requires one.